#  <p style="text-align: center;">Практическое задание к Семинару 6.<br><br>Урок 6. Сравнение долей. Построение доверительного интервала.</p>

In [17]:
import numpy as np
import statistics as st
import scipy.stats as stats
import math

**Z-критерий**
<div class="alert alert-block alert-warning"><font size=4 color=black><b>$\displaystyle\bar X\pm z_{\alpha/2}\cdot\frac{\sigma}{\sqrt n}$, (1) </b></font></div>где: <b>X</b> - среднее значение выборки, $z_{\alpha/2}$ табличное значение <b>Z</b>-критерия, 
$\sigma$-среднее квадратическое отклонение генеральной совокупности, <b>n</b> - объем выборки

**Критерий Стьюдента (t)**
<div class="alert alert-block alert-warning"><font size=4 color=black><b>$\displaystyle\bar X\pm t_{\alpha/2}\cdot\frac{\sigma_{несмещ}}{\sqrt n}$, (2) </b></font></div>где: <b>X</b> - среднее значение выборки, $t_{\alpha/2}$ - табличное значение критерия Стьюдента, 
$\sigma$ - несмещённое среднее квадратическое отклонение выборки, <b>n</b> - объем выборки

Таблица 1. Z Таблица (положительные значения)

![Positive Z Table](https://datascience.eu/wp-content/uploads/2020/05/image1.png)

Таблица 2. Z Таблица (отрицательные значения)

![Negative Z Table](https://www.ztable.net/wp-content/uploads/2018/11/negativeztable.png)

Таблица 3. Таблица Стьюдента   

![Student Table](https://fsd.multiurok.ru/viewImage.php?image=http://nice-diplom.ru/templates/blue/images/img/form_69.gif)

In [18]:
def Z_interval (X_mean, alpha, sigma, n):
    Z_table = stats.norm.ppf(1-alpha/2) # Также можно найти по Таблице 1
    print (f'[{X_mean - (Z_table*(sigma/(n**0.5)))}, {X_mean + (Z_table*(sigma/(n**0.5)))}]')

In [19]:
def t_interval (array, alpha):
    array_mean = np.mean(array)
    D = np.var(array, ddof=1)
    t = stats.t.ppf(1-alpha/2, len(array) - 1) # Также можно найти по Таблице 2
    print (f'[{array_mean - t * np.sqrt(D/len(array))}, {array_mean + t * np.sqrt(D/len(array))}]')

In [20]:
def interval_for_mean (array_1, array_2, alpha):     # Интервальная оценка для разности средних арифметических
    raznost = array_1 - array_2
    raznost_mean = np.mean(raznost)
    n = len(raznost)
    raznost_var = raznost.var()*n/(n-1)
    sigma = np.sqrt(raznost_var)
    t_kriteriy = stats.t.ppf((1+(1-alpha))/2, n-1)
    print (f'[{raznost_mean - t_kriteriy * sigma/np.sqrt(n)}, {raznost_mean + t_kriteriy * sigma/np.sqrt(n)}]')

## ЗАДАНИЕ 1. 
**Известно, что генеральная совокупность распределена нормально со средним квадратическим отклонением, равным 16. Найти доверительный интервал для оценки математического ожидания a с надежностью 0.95, если выборочная средняя M = 80, а объем выборки n = 256.**

**<p style="text-align: center;">
    <font size=10 color=#32CD32><b>-------------------------------</b></font><br>
    <font size=4.5 color=#3ac444>РЕШЕНИЕ:</font><br>
<font size=10 color=#32CD32><b>-------------------------------</b></font></p>**

Т.к., из условия задачи нам известно среднее квадратическое отклонение генеральной совокупности $\sigma=16$, то для расчета $95\%$-го доверительного интервала, будем использовать $Z$-критерий. Границы доверительного интервала расчитаем по формуле (1) и будет равен:<br>
Левая граница: $80-1.96\cdot\frac{16}{\sqrt 256}$<br>
Правая граница: $80+1.96\cdot\frac{16}{\sqrt 256}$

In [21]:
# Дано
sigma = 16
alpha = 0.05
n = 256
X_mean = 80
Z_interval (X_mean, alpha, sigma, n)

[78.04003601545995, 81.95996398454005]


<div class="alert alert-block alert-success"><font size=3 color='b'><b>ОТВЕТ: </b><br>Доверительный интервал для оценки математического ожидания с надежностью 0.95? будет находиться в границах [78.04, 81.96]</font></div>

---

## ЗАДАНИЕ 2. 
**В результате 10 независимых измерений некоторой величины X, выполненных с одинаковой точностью, получены опытные данные:**
```
6.9, 6.1, 6.2, 6.8, 7.5, 6.3, 6.4, 6.9, 6.7, 6.1
```
**Предполагая, что результаты измерений подчинены нормальному закону распределения вероятностей, оценить истинное значение величины X при помощи доверительного интервала, покрывающего это значение с доверительной вероятностью 0,95.**

**<p style="text-align: center;">
    <font size=10 color=#32CD32><b>-------------------------------</b></font><br>
    <font size=4.5 color=#3ac444>РЕШЕНИЕ:</font><br>
<font size=10 color=#32CD32><b>-------------------------------</b></font></p>**

Т.к., из условия задачи нам НЕ известно мат. ожидании генеральной совокупности, СКО для генеральной совокупности, то для расчета $95\%$-го доверительного интервала будем использовать $t$-критерий и формулу (2).

In [22]:
# Дано
data = [6.9, 6.1, 6.2, 6.8, 7.5, 6.3, 6.4, 6.9, 6.7, 6.1]
data_mean = np.mean(data)
alpha = 0.05
n = 10
t_interval (data, alpha)

[6.267515851415713, 6.912484148584288]


<div class="alert alert-block alert-success"><font size=3 color='b'><b>ОТВЕТ: </b><br>Истинное значение величины X будет находиться в интервале [6.268, 6.912]</font></div>

---

## ЗАДАНИЕ 3. 
*Рост дочерей:*
```
175, 167, 154, 174, 178, 148, 160, 167, 169, 170
``` 
*Рост матерей:*
```
178, 165, 165, 173, 168, 155, 160, 164, 178, 175
```
**Используя эти данные построить 95% доверительный интервал для разности среднего роста родителей и детей.**

**<p style="text-align: center;">
    <font size=10 color=#32CD32><b>-------------------------------</b></font><br>
    <font size=4.5 color=#3ac444>РЕШЕНИЕ:</font><br>
<font size=10 color=#32CD32><b>-------------------------------</b></font></p>**

Т.к., из условия задачи нам НЕ известно мат. ожидании генеральной совокупности, СКО для генеральной совокупности, то для расчета $95\%$-го доверительного интервала будем использовать $t$-критерий и формулу (2).

In [23]:
daughters = np.array([175, 167, 154, 174, 178, 148, 160, 167, 169, 170])
mothers = np.array([178, 165, 165, 173, 168, 155, 160, 164, 178, 175])
delta = np.mean(daughters) - np.mean(mothers)
D1 = np.var(mothers, ddof=1)
D2 = np.var(daughters, ddof=1)
D = (D1 + D2) /2
SE = np.sqrt(D/len(mothers) + D/len(daughters)) #Ошибка разницы средних
t = stats.t.ppf(0.975, 2*len(mothers) - 2) 
left = raznost_means - t * SE # Нижняя (левая) граница интервала
right = raznost_means + t * SE # Верхняя (правая) граница интервала
print(f'95% доверительный интервал для разности среднего роста родителей и детей: [{left:.2f}, {right:.2f}]')

95% доверительный интервал для разности среднего роста родителей и детей: [-6.27, 10.07]


<div class="alert alert-block alert-success"><font size=3 color='b'><b>ОТВЕТ: </b><br>95% lоверительный интервал для разности среднего роста РОДИТЕЛЕЙ и ДЕТЕЙ [-6.27, 10.07]</font></div>